src/mallocTracer.cpp

```diff
@@ -29,17 +29,33 @@ static realloc_t _orig_realloc = NULL;
 typedef void (*free_t)(void*);
 static free_t _orig_free = NULL;
 
+typedef int (*posix_memalign_t)(void**, size_t, size_t);
+static posix_memalign_t _orig_posix_memalign = NULL;
+
+typedef void* (*aligned_alloc_t)(size_t, size_t);
+static aligned_alloc_t _orig_aligned_alloc = NULL;
+
+// In musl, posix_memalign calls aligned_alloc and it in turns calls malloc.
+// If in_musl_posix_memalign, do not recordMalloc in aligned_alloc_hook.
+// If in_musl_aligned_alloc, do not recordMalloc in malloc_hook.
+// Do not access thread_local variables in non-musl.
+static bool is_musl = false;
+static thread_local bool in_musl_posix_memalign = false;
```

Comment:
We hit several issues with deadlocks and thread locals.
Our solution was to control the [creation and deletion of the thread local](https://github.com/DataDog/ddprof/blob/cf9ec330672eb0cf597a89559ed5948f940ae416/src/lib/allocation_tracker.cc#L41) variables with get/set_specific.
That implies hooking thread creation to ensure you can create these states. Though I think you already have such hooks.
We also added guards to avoid re-entrance (this avoids double counting mentioned bellow).
I'm not sure if this is preferable to your approach, though feel free to compare.

Reply:
Thanks, we are looking into a simpler implementation that does *not* avoid double-accounting, yet still has correct leak detection.